### PR TITLE
Add download-logs command to gei and ado2gh

### PR DESCRIPTION
**Pull request by synthead**
Depends on https://github.com/github/octoshift/pull/4521
Depends on https://github.com/github/octoshift/pull/4520
Depends on https://github.com/github/github/pull/219142
Part of https://github.com/github/octoshift/issues/4380

Adds support for downloading migration logs for migrations.

- [x] Did you write/update appropriate tests
- [x] Release notes updated (if appropriate)
- [x] Appropriate logging output
- [x] Issue linked
- [x] Docs updated (or issue created) (https://github.com/github/gh-gei/issues/381)

### CODE REVIEW

File: src/Octoshift/GithubApi.cs [hunk 1]
```diff
@@ -341,6 +341,34 @@ public virtual async Task<string> GetMigrationFailureReason(string migrationId)
             return (string)data["data"]["node"]["failureReason"];
         }
 
+        public virtual async Task<string> GetMigrationLogUrl(string orgLogin, string repositoryName)
```
Comment (dylan-smith) [inline src/Octoshift/GithubApi.cs:344]: every other function in here uses the arg name "org" not org login. can we make it consistent.
```suggestion
        public virtual async Task<string> GetMigrationLogUrl(string org, string repositoryName)
```

File: src/Octoshift/GithubApi.cs [hunk 2]
```diff
@@ -341,6 +341,34 @@ public virtual async Task<string> GetMigrationFailureReason(string migrationId)
...
+                }
+              }
+            ";
+
+            var payload = new { query = $"{query} {{ {gql} }}", variables = new { orgLogin, repositoryName } };
+
+            var response = await _retryPolicy.Retry(
+              async () => await _client.PostAsync(url, payload),
+              ex => ex.StatusCode == HttpStatusCode.BadGateway
+            );
```
Comment (dylan-smith) [inline src/Octoshift/GithubApi.cs:364]: Do we expect this to return BadGateway errors in some circumstances?

File: src/Octoshift/HttpDownloadService.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    {
+        internal Func<string, string, Task> WriteToFile = async (path, contents) => await File.WriteAllTextAsync(path, contents);
+
+        private readonly HttpClient _httpClient;
+
+        public HttpDownloadService(HttpClient httpClient) => _httpClient = httpClient;
+
+        public virtual async Task Download(string url, string file)
+        {
+            using var response = await _httpClient.GetAsync(url, HttpCompletionOption.ResponseHeadersRead);
```
Comment (dylan-smith) [inline src/Octoshift/HttpDownloadService.cs:20]: I don't think we need the 2nd argument here. Unless you put it here for a specific reason. It looks like GetAsync(url) would work to me.

File: src/gei/Commands/DownloadLogsCommand.cs [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            var targetRepo = new Option<string>("--target-repo")
+            {
+                IsRequired = true,
+                Description = "Target repository to download latest log for."
+            };
+
+            var targetApiUrl = new Option<string>("--target-api-url")
+            {
+                IsRequired = false,
+                Description = "Target GitHub API URL if not migrating to github.com (default: https://api.github.com)."
```
Comment (dylan-smith) [inline src/gei/Commands/DownloadLogsCommand.cs:47]:
```suggestion
                Description = "Target GitHub API URL if not targeting github.com (default: https://api.github.com)."
```

**Review by dylan-smith — CHANGES_REQUESTED** [review]

File: src/Octoshift/GithubApi.cs (hunk 1 above)
Comment (synthead) [reply src/Octoshift/GithubApi.cs]: Updated in 969e878dcc0f15a5fd2569cb296ca01f979dc6ef!

File: src/Octoshift/GithubApi.cs (hunk 2 above)
Comment (synthead) [reply src/Octoshift/GithubApi.cs:364]: I copied prior art from other functions in this class, i.e.:

https://github.com/github/gh-gei/blob/969e878dcc0f15a5fd2569cb296ca01f979dc6ef/src/Octoshift/GithubApi.cs#L278-L279

Would you like me to remove exception handling for BadGateway?

File: src/Octoshift/HttpDownloadService.cs (hunk 3 above)
Comment (synthead) [reply src/Octoshift/HttpDownloadService.cs:20]: Good call! Removed in 88f1a5efaa525a9b118c08df30df99b4435ce70e!

Comment (synthead) [reply src/Octoshift/HttpDownloadService.cs]: Good call!  Removed in 88f1a5efaa525a9b118c08df30df99b4435ce70e!

File: src/Octoshift/GithubApi.cs (hunk 2 above)
Comment (synthead) [reply src/Octoshift/GithubApi.cs]: You got it!  Removed in d574f5872c05c71d022eb5ce395f1a9833f8f115!

File: src/gei/Commands/DownloadLogsCommand.cs (hunk 4 above)
Comment (synthead) [reply src/gei/Commands/DownloadLogsCommand.cs]: Updated with 3c902c89b63335e9b1a485f8767ea2aa8f124cca!

**Review by dylan-smith — APPROVED** [review]
:shipit:

**Pull request merged**